### ISSUE TITLE
classId interferes with snapshot testing

### ISSUE THREAD

**Issue by jistjoalal** [timhagn/gatsby-background-image#98]
## Summary

The unique classId that is being deprecated is a minor annoyance for snapshot testing with Jest.

## Relevant information

The unique classId causes snapshot testing to fail:
![tests](https://i.gyazo.com/8bd0acf78fbe1e98fbb7470b77fd5880.png)

When classId is manually set to avoid failing snapshots, I get a bunch of warnings about it being deprecated:
![warning](https://i.gyazo.com/7864d16bfc746abd2643516d89ebb015.png)

That is all I could fit on my screen. This is the output from one test of a simple footer component which uses the BackgroundImage component only once.

## Question

Is there a way to silence these warnings or another way of getting this component to not break snapshot testing?

**Comment by timhagn** [comment]
Hi @jistjoalal,

gonna look into that when I'm back home. You might have found the first real use case for `classId` % ). Might be a hard one to go at (unless not deprecating it), as I wouldn't know of a way to reuse classes in tests.
For a start I'm gonna remove the depreciation warning.
Maybe we could deactivate the unique class generation in `env=test` cases, though this might only be possible for cases like yours, multiple `BackgroundImage` components on one page would then break tests % |...
Any ideas?

Best,

Tim.

**Comment by timhagn** [comment]
Ah, ok, just found [property matchers](https://jestjs.io/docs/en/snapshot-testing#property-matchers) for snapshot testing, so `classId` can still move to the bin : ).

**Comment by jistjoalal** [comment]
Thanks for the help, Tim.

Unfortunately because BackgroundImage is quite a complex object with the generated class name appearing multiple times, it seems like it would be difficult to write a property matcher.

I thought this would be enough to get Jest to ignore it:
```js
    expect(tree).toMatchSnapshot({
      props: {
        className: expect.any(String),
      },
    })
```
But because of the nested style blocks it still fails snapshot testing:
![nested styles](https://i.gyazo.com/49889edfed6c76f261a0924cb3550061.png)

Maybe I'm missing something, but as I currently understand it I think snapshot testing with Jest is just not a good solution for me. I use BackgroundImage as a part of my site layout and so would need to write this complex property matcher for every snapshot I take. My goal was simply to have a way of protecting against visual regressions while refactoring my code (without clicking through every page).

I'll close the issue because I think I've found a better solution to my problem. [Cypress' Visual Testing](https://docs.cypress.io/guides/tooling/visual-testing.html#Functional-vs-visual-testing) seems to be a much better fit for what I'm trying to accomplish. With something like this I can test my site the way I need without having to mock a bunch of stuff or worry about implementation details of my dependencies.

I won't waste any more of your time, but if you've got any thoughts for a testing noob that'd be awesome!

**Comment by timhagn** [comment]
Hi @jistjoalal,

darn, completely forgot about the inline styling! Of course you're completely right, they'd fail nonetheless. Cypress seems a good solution for your case, already experimented with it a little and am quite impressed by the smoothness of the integration. 
Let me know how the experience is for you : )!
Shouldn't you want to be dependent on a cloud solution, be sure to have a look at [pupeteer](https://github.com/puppeteer/puppeteer/) or (it's quasi "successor", if one believes the hype) [playwright](https://github.com/microsoft/playwright), which automate "UI" tests in a local headless browser.

And you didn't waste my time but are quite welcome for bringing this up, as it might bite `gbi` in the future (and brought my thoughts to some advancements for v1 (or for same rather v2 ; ).

Happy testing : )!

Best,

Tim.

**Comment by ebello** [comment]
Hi, just wanted to bring up this use case again. We use snapshot testing extensively and having the class name change every render is causing our tests to fail. Is there a workaround other than using the deprecated `classId`?

**Comment by timhagn** [comment]
Hi @ebello & @jistjoalal,

how could I have forgotten that the solution already was in my test suite for `gbi`! 
Definitely gonna add a "Testing `gbi`" paragraph in the next README, but now straight to it % ).

As `gbi` uses `short-uuid` to create it's unique classes, you only have to mock its `generate()` function like I'm gonna explain now.

Either in your `jest.setup.js` or the top of your individual test file(s) mock the complete package:
`jest.mock('short-uuid')`

Then for each `gbi` component you want to test, add a `beforeEach()`:

```js
beforeEach(() => {
    // Freeze the generated className.
    const uuid = require('short-uuid')
    uuid.generate.mockImplementation(() => '73WakrfVbNJBaAmhQtEeDv')
});
```

Now the class name will always be the same and your snapshot tests should work : ).

Gonna add a snapshot test to `gbitest` later with the aforementioned mocking - so I hope we can finally flush the additional classId down the drain of history ; ).

Happy testing!

Best,

Tim.

**Comment by timhagn** [comment]
Hi again @ebello & @jistjoalal,

as promised: with the just released `v1.0.0` of `gbi` (removing deprecated props) `gbitest` has now a Demo-Snapshot test : ). Hope you are able to switch to it with the added [testing information](https://github.com/timhagn/gatsby-background-image/tree/master/packages/gatsby-background-image#testing-gatsby-background-image) ; ).

Best,

Tim.
